cpp/src/arrow/compute/kernels/codegen_internal.cc

```diff
@@ -282,15 +296,31 @@ TypeHolder CommonTemporal(const TypeHolder* begin, size_t count) {
     }
   }
 
-  if (timezone) {
-    // At least one timestamp seen
-    return timestamp(finest_unit, *timezone);
-  } else if (saw_date64) {
-    return date64();
-  } else if (saw_date32) {
-    return date32();
-  } else if (saw_duration) {
-    return duration(finest_unit);
+  bool has_saw_time_since_midnight = saw_time32 || saw_time64;
+  bool has_saw_time_or_date = timezone || saw_date64 || saw_date32 || saw_duration;
+
+  if (has_saw_time_since_midnight && has_saw_time_or_date) {
```

Comment:
I think time32/time64 means time in a day, so they can't cast to timestamp and related type

Reply:
That's correct. It's not supported arithmetic currently, but I think we could eventually define `time + date -> timestamp`

Reply:
Yeah, right, so "add" should handle the type more carefully?